File: TI1/TI1.3 RESTFUL_JAVA_SINBDD_GR09/01.SERVIDOR/CONUNI_RESTFUL_JAVA_GR09/src/main/java/ec/edu/monster/servicios/ConversionService.java
package ec.edu.monster.servicios;

import ec.edu.monster.modelo.Conversion;
import ec.edu.monster.modelo.ConversionResponse;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Named;

/**
 * Servicio para conversiones de temperatura y unidades
 * Capa de servicios que contiene toda la lógica de negocio
 * @author ACER NITRO V15
 */
@Named
@ApplicationScoped
public class ConversionService {
    
    // ========== CONVERSIONES DE TEMPERATURA ==========
    
    /**
     * Convierte Celsius a Fahrenheit
     */
    public double celsiusAFahrenheit(double celsius) {
        return (celsius * Conversion.CELSIUS_TO_FAHRENHEIT_MULTIPLIER) + Conversion.CELSIUS_TO_FAHRENHEIT_OFFSET;
    }
    
    /**
     * Convierte Fahrenheit a Celsius
     */
    public double fahrenheitACelsius(double fahrenheit) {
        return (fahrenheit - Conversion.CELSIUS_TO_FAHRENHEIT_OFFSET) / Conversion.CELSIUS_TO_FAHRENHEIT_MULTIPLIER;
    }
    
    /**
     * Convierte Celsius a Kelvin
     */
    public double celsiusAKelvin(double celsius) {
        return celsius + Conversion.KELVIN_OFFSET;
    }
    
    /**
     * Convierte Kelvin a Celsius
     */
    public double kelvinACelsius(double kelvin) {
        return kelvin - Conversion.KELVIN_OFFSET;
    }
    
    /**
     * Convierte Fahrenheit a Kelvin
     */
    public double fahrenheitAKelvin(double fahrenheit) {
        return fahrenheitACelsius(fahrenheit) + Conversion.KELVIN_OFFSET;
    }
    
    /**
     * Convierte Kelvin a Fahrenheit
     */
    public double kelvinAFahrenheit(double kelvin) {
        return celsiusAFahrenheit(kelvinACelsius(kelvin));
    }
    
    // ========== CONVERSIONES DE LONGITUD ==========
    
    /**
     * Convierte metros a pies
     */
    public double metrosAPies(double metros) {
        return metros * Conversion.METERS_TO_FEET;
    }
    
    /**
     * Convierte pies a metros
     */
    public double piesAMetros(double pies) {
        return pies / Conversion.METERS_TO_FEET;
    }
    
    /**
     * Convierte metros a pulgadas
     */
    public double metrosAPulgadas(double metros) {
        return metros * Conversion.METERS_TO_INCHES;
    }
    
    /**
     * Convierte pulgadas a metros
     */
    public double pulgadasAMetros(double pulgadas) {
        return pulgadas / Conversion.METERS_TO_INCHES;
    }
    
    /**
     * Convierte kilómetros a millas
     */
    public double kilometrosAMillas(double kilometros) {
        return kilometros * Conversion.KILOMETERS_TO_MILES;
    }
    
    /**
     * Convierte millas a kilómetros
     */
    public double millasAKilometros(double millas) {
        return millas / Conversion.KILOMETERS_TO_MILES;
    }
    
    // ========== CONVERSIONES DE PESO/MASA ==========
    
    /**
     * Convierte kilogramos a libras
     */
    public double kilogramosALibras(double kilogramos) {
        return kilogramos * Conversion.KILOGRAMS_TO_POUNDS;
    }
    
    /**
     * Convierte libras a kilogramos
     */
    public double librasAKilogramos(double libras) {
        return libras / Conversion.KILOGRAMS_TO_POUNDS;
    }
    
    /**
     * Convierte gramos a onzas
     */
    public double gramosAOnzas(double gramos) {
        return gramos * Conversion.GRAMS_TO_OUNCES;
    }
    
    /**
     * Convierte onzas a gramos
     */
    public double onzasAGramos(double onzas) {
        return onzas / Conversion.GRAMS_TO_OUNCES;
    }
    
    // ========== CONVERSIONES DE VOLUMEN ==========
    
    /**
     * Convierte litros a galones
     */
    public double litrosAGalones(double litros) {
        return litros * Conversion.LITERS_TO_GALLONS;
    }
    
    /**
     * Convierte galones a litros
     */
    public double galonesALitros(double galones) {
        return galones / Conversion.LITERS_TO_GALLONS;
    }
    
    /**
     * Convierte mililitros a onzas fluidas
     */
    public double mililitrosAOnzasFluidas(double mililitros) {
        return mililitros * Conversion.MILLILITERS_TO_FLUID_OUNCES;
    }
    
    /**
     * Convierte onzas fluidas a mililitros
     */
    public double onzasFluidasAMililitros(double onzasFluidas) {
        return onzasFluidas / Conversion.MILLILITERS_TO_FLUID_OUNCES;
    }
    
    // ========== CONVERSIONES DE ÁREA ==========
    
    /**
     * Convierte metros cuadrados a pies cuadrados
     */
    public double metrosCuadradosAPiesCuadrados(double metrosCuadrados) {
        return metrosCuadrados * Conversion.SQUARE_METERS_TO_SQUARE_FEET;
    }
    
    /**
     * Convierte pies cuadrados a metros cuadrados
     */
    public double piesCuadradosAMetrosCuadrados(double piesCuadrados) {
        return piesCuadrados / Conversion.SQUARE_METERS_TO_SQUARE_FEET;
    }
    
    /**
     * Convierte hectáreas a acres
     */
    public double hectareasAAcres(double hectareas) {
        return hectareas * Conversion.HECTARES_TO_ACRES;
    }
    
    /**
     * Convierte acres a hectáreas
     */
    public double acresAHectareas(double acres) {
        return acres / Conversion.HECTARES_TO_ACRES;
    }
    
    // ========== MÉTODO PRINCIPAL DE CONVERSIÓN ==========
    
    /**
     * Método principal que maneja todas las conversiones
     * @param valor Valor a convertir
     * @param unidadOrigen Unidad de origen
     * @param unidadDestino Unidad de destino
     * @param categoria Categoría de conversión
     * @return ConversionResponse con el resultado
     */
    public ConversionResponse convertir(double valor, String unidadOrigen, String unidadDestino, String categoria) {
        try {
            // Validar entrada
            if (valor < 0 && !categoria.equals(Conversion.CATEGORIA_TEMPERATURA)) {
                return ConversionResponse.crearError("El valor no puede ser negativo para esta categoría");
            }
            
            if (!Conversion.esConversionValida(unidadOrigen, unidadDestino, categoria)) {
                return ConversionResponse.crearError("Conversión no válida entre " + unidadOrigen + " y " + unidadDestino);
            }
            
            double resultado = realizarConversion(valor, unidadOrigen, unidadDestino, categoria);
            
            return ConversionResponse.crearExito(valor, resultado, unidadOrigen, unidadDestino, categoria);
            
        } catch (Exception e) {
            return ConversionResponse.crearError("Error en la conversión: " + e.getMessage());
        }
    }
    
    /**
     * Realiza la conversión específica según los parámetros
     */
    private double realizarConversion(double valor, String unidadOrigen, String unidadDestino, String categoria) {
        String operacion = unidadOrigen + "A" + unidadDestino;
        
        switch (categoria.toLowerCase()) {
            case Conversion.CATEGORIA_TEMPERATURA:
                return convertirTemperatura(valor, operacion);
            case Conversion.CATEGORIA_LONGITUD:
                return convertirLongitud(valor, operacion);
            case Conversion.CATEGORIA_PESO:
                return convertirPeso(valor, operacion);
            case Conversion.CATEGORIA_VOLUMEN:
                return convertirVolumen(valor, operacion);
            case Conversion.CATEGORIA_AREA:
                return convertirArea(valor, operacion);
            default:
                throw new IllegalArgumentException("Categoría no soportada: " + categoria);
        }
    }
    
    private double convertirTemperatura(double valor, String operacion) {
        switch (operacion) {
            case "celsiusAfahrenheit": return celsiusAFahrenheit(valor);
            case "fahrenheitAcelsius": return fahrenheitACelsius(valor);
            case "celsiusAkelvin": return celsiusAKelvin(valor);
            case "kelvinAcelsius": return kelvinACelsius(valor);
            case "fahrenheitAkelvin": return fahrenheitAKelvin(valor);
            case "kelvinAfahrenheit": return kelvinAFahrenheit(valor);
            default: throw new IllegalArgumentException("Operación de temperatura no soportada: " + operacion);
        }
    }
    
    private double convertirLongitud(double valor, String operacion) {
        switch (operacion) {
            case "metrosApies": return metrosAPies(valor);
            case "piesAmetros": return piesAMetros(valor);
            case "metrosApulgadas": return metrosAPulgadas(valor);
            case "pulgadasAmetros": return pulgadasAMetros(valor);
            case "kilometrosAmillas": return kilometrosAMillas(valor);
            case "millasAkilometros": return millasAKilometros(valor);
            default: throw new IllegalArgumentException("Operación de longitud no soportada: " + operacion);
        }
    }
    
    private double convertirPeso(double valor, String operacion) {
        switch (operacion) {
            case "kilogramosAlibras": return kilogramosALibras(valor);
            case "librasAkilogramos": return librasAKilogramos(valor);
            case "gramosAonzas": return gramosAOnzas(valor);
            case "onzasAgramos": return onzasAGramos(valor);
            default: throw new IllegalArgumentException("Operación de peso no soportada: " + operacion);
        }
    }
    
    private double convertirVolumen(double valor, String operacion) {
        switch (operacion) {
            case "litrosAgalones": return litrosAGalones(valor);
            case "galonesAlitros": return galonesALitros(valor);
            case "mililitrosAonzasFluidas": return mililitrosAOnzasFluidas(valor);
            case "onzasFluidasAmililitros": return onzasFluidasAMililitros(valor);
            default: throw new IllegalArgumentException("Operación de volumen no soportada: " + operacion);
        }
    }
    
    private double convertirArea(double valor, String operacion) {
        switch (operacion) {
            case "metrosCuadradosApiesCuadrados": return metrosCuadradosAPiesCuadrados(valor);
            case "piesCuadradosAmetrosCuadrados": return piesCuadradosAMetrosCuadrados(valor);
            case "hectareasAacres": return hectareasAAcres(valor);
            case "acresAhectareas": return acresAHectareas(valor);
            default: throw new IllegalArgumentException("Operación de área no soportada: " + operacion);
        }
    }
}

File: TI1/TI1.4 RESTFUL_DOTNET_SINBDD_GR09/01.SERVIDOR/CONUNI_RESTFUL_DOTNET_GR9S/CONUNI_RESTFUL_DOTNET_GR9S/Models/ConversionService.cs
using System;

namespace CONUNI_RESTFUL_DOTNET_GR9S.Models
{
    /// <summary>
    /// Servicio de lógica de negocio para conversiones.
    /// </summary>
    public class ConversionService
    {
        public double CelsiusAFahrenheit(double celsius)
        {
            return (celsius * Conversion.CELSIUS_TO_FAHRENHEIT_MULTIPLIER) + Conversion.CELSIUS_TO_FAHRENHEIT_OFFSET;
        }

        public double FahrenheitACelsius(double fahrenheit)
        {
            return (fahrenheit - Conversion.CELSIUS_TO_FAHRENHEIT_OFFSET) / Conversion.CELSIUS_TO_FAHRENHEIT_MULTIPLIER;
        }

        public double CelsiusAKelvin(double celsius)
        {
            return celsius + Conversion.KELVIN_OFFSET;
        }

        public double KelvinACelsius(double kelvin)
        {
            return kelvin - Conversion.KELVIN_OFFSET;
        }

        public double FahrenheitAKelvin(double fahrenheit)
        {
            return CelsiusAKelvin(FahrenheitACelsius(fahrenheit));
        }

        public double KelvinAFahrenheit(double kelvin)
        {
            return CelsiusAFahrenheit(KelvinACelsius(kelvin));
        }

        public double MetrosAPies(double metros)
        {
            return metros * Conversion.METERS_TO_FEET;
        }

        public double PiesAMetros(double pies)
        {
            return pies / Conversion.METERS_TO_FEET;
        }

        public double MetrosAPulgadas(double metros)
        {
            return metros * Conversion.METERS_TO_INCHES;
        }

        public double PulgadasAMetros(double pulgadas)
        {
            return pulgadas / Conversion.METERS_TO_INCHES;
        }

        public double KilometrosAMillas(double kilometros)
        {
            return kilometros * Conversion.KILOMETERS_TO_MILES;
        }

        public double MillasAKilometros(double millas)
        {
            return millas / Conversion.KILOMETERS_TO_MILES;
        }

        public double KilogramosALibras(double kilogramos)
        {
            return kilogramos * Conversion.KILOGRAMS_TO_POUNDS;
        }

        public double LibrasAKilogramos(double libras)
        {
            return libras / Conversion.KILOGRAMS_TO_POUNDS;
        }

        public double GramosAOnzas(double gramos)
        {
            return gramos * Conversion.GRAMS_TO_OUNCES;
        }

        public double OnzasAGramos(double onzas)
        {
            return onzas / Conversion.GRAMS_TO_OUNCES;
        }

        public ConversionResponse Convertir(double valor, string unidadOrigen, string unidadDestino, string categoria)
        {
            try
            {
                if (valor < 0 && !Conversion.CATEGORIA_TEMPERATURA.Equals(categoria, StringComparison.OrdinalIgnoreCase))
                {
                    return ConversionResponse.CrearError("El valor no puede ser negativo para esta categoría");
                }

                if (!Conversion.EsConversionValida(unidadOrigen, unidadDestino, categoria))
                {
                    return ConversionResponse.CrearError($"Conversión no válida entre {unidadOrigen} y {unidadDestino}");
                }

                double resultado = RealizarConversion(valor, unidadOrigen, unidadDestino, categoria);
                return ConversionResponse.CrearExito(valor, resultado, unidadOrigen, unidadDestino, categoria);
            }
            catch (Exception ex)
            {
                return ConversionResponse.CrearError($"Error en la conversión: {ex.Message}");
            }
        }

        private double RealizarConversion(double valor, string unidadOrigen, string unidadDestino, string categoria)
        {
            string origen = unidadOrigen.ToLowerInvariant();
            string destino = unidadDestino.ToLowerInvariant();

            switch (categoria.ToLowerInvariant())
            {
                case Conversion.CATEGORIA_TEMPERATURA:
                    return ConvertirTemperatura(valor, origen, destino);
                case Conversion.CATEGORIA_LONGITUD:
                    return ConvertirLongitud(valor, origen, destino);
                case Conversion.CATEGORIA_PESO:
                    return ConvertirPeso(valor, origen, destino);
                default:
                    throw new ArgumentException($"Categoría no soportada: {categoria}");
            }
        }

        private double ConvertirTemperatura(double valor, string origen, string destino)
        {
            switch ($"{origen}A{destino}")
            {
                case "celsiusAfahrenheit":
                    return CelsiusAFahrenheit(valor);
                case "fahrenheitAcelsius":
                    return FahrenheitACelsius(valor);
                case "celsiusAkelvin":
                    return CelsiusAKelvin(valor);
                case "kelvinAcelsius":
                    return KelvinACelsius(valor);
                case "fahrenheitAkelvin":
                    return FahrenheitAKelvin(valor);
                case "kelvinAfahrenheit":
                    return KelvinAFahrenheit(valor);
                default:
                    throw new ArgumentException($"Operación de temperatura no soportada: {origen} -> {destino}");
            }
        }

        private double ConvertirLongitud(double valor, string origen, string destino)
        {
            switch ($"{origen}A{destino}")
            {
                case "metrosApies":
                    return MetrosAPies(valor);
                case "piesAmetros":
                    return PiesAMetros(valor);
                case "metrosApulgadas":
                    return MetrosAPulgadas(valor);
                case "pulgadasAmetros":
                    return PulgadasAMetros(valor);
                case "kilometrosAmillas":
                    return KilometrosAMillas(valor);
                case "millasAkilometros":
                    return MillasAKilometros(valor);
                default:
                    throw new ArgumentException($"Operación de longitud no soportada: {origen} -> {destino}");
            }
        }

        private double ConvertirPeso(double valor, string origen, string destino)
        {
            switch ($"{origen}A{destino}")
            {
                case "kilogramosAlibras":
                    return KilogramosALibras(valor);
                case "librasAkilogramos":
                    return LibrasAKilogramos(valor);
                case "gramosAonzas":
                    return GramosAOnzas(valor);
                case "onzasAgramos":
                    return OnzasAGramos(valor);
                default:
                    throw new ArgumentException($"Operación de peso no soportada: {origen} -> {destino}");
            }
        }
    }
}
